autoflow/autoflow/parser.py

```diff
@@ -0,0 +1,450 @@
+    to a DateStencil object.
+    """
+
+    def _deserialize(self, value, attr, data, **kwargs) -> DateStencil:
+        """
+        Create a DateStencil object from the input value.
+
+        Returns
+        -------
+        DateStencil
+
+        Raises
+        ------
+        ValidationError
+            if the input value is not a valid 'raw' stencil
+
+        See also
+        --------
+        autoflow.date_stencil.DateStencil
+        """
+        try:
+            return DateStencil(raw_stencil=value)
+        except (TypeError, ValueError, InvalidDateIntervalError) as e:
+            raise ValidationError(str(e))
+
+
+class NotebookOutputSchema(Schema):
+    """
+    Schema for the 'output' parameter of a notebook task specification.
+
+    Fields
+    ------
+    format : str
+        Format to which notebook should be converted (only 'pdf' is currently supported).
+    template : str, optional
+        Filename of a template file to be used when converting the notebook to ASCIIDoc format.
+        If not provided, deserialises to None.
+    """
+
+    format = fields.String(validate=validate.OneOf(["pdf"]), required=True)
+    template = fields.String(missing=None)
+
+    @validates("template")
+    def validate_template(self, value):
+        """
+        Raise a ValidationError if the specified template file does not exist,
+        or if the inputs directory is not specified in the context.
+        """
+        if value is None:
+            return
+        try:
+            inputs_dir = Path(self.context["inputs_dir"])
+        except KeyError:
+            raise ValidationError(
+                "'inputs_dir' was not provided in the context. Cannot check for file existence."
+            )
+        if not (inputs_dir / value).exists():
+            raise ValidationError(f"Asciidoc template file '{value}' not found.")
+
+
+class NotebookSchema(Schema):
+    """
+    Schema for a notebook task specification.
+
+    Fields
+    ------
+    filename : str
+        Filename of Jupyter notebook to be executed.
+    parameters : dict, optional
+        Dictionary mapping parameter names (as they will be accessible within
+        the notebook) to the names of task outputs within the surrounding workflow.
+    output : Nested(NotebookOutputSchema), optional
+        Configuration parameters for the output generation task.
+    """
+
+    filename = fields.String(required=True)
+    parameters = fields.Dict(
+        keys=fields.String(), values=fields.String(), required=False
+    )
+    output = fields.Nested(NotebookOutputSchema, required=False)
+
+    @validates("filename")
+    def validate_filename(self, value):
+        """
+        Raise a ValidationError if the specified file does not exist or is not a Jupyter notebook,
+        or if the inputs directory is not specified in the context.
+        """
+        if Path(value).suffix != ".ipynb":
+            # TODO: '.ipynb' does not a notebook make
```

Comment:
Not a big deal to my mind - you could validate that it is a true notebook by parsing the json, but that's a _lot_ of effort that doesn't really gain you anything.